### ISSUE TITLE
System.Runtime.InteropServices.ExternalException : A generic error occurred in GDI+

### ISSUE THREAD

**Issue by StefanKoell** [dotnet/runtime#58741]
* .NET Core Version: 
Microsoft.WindowsDesktop.App 5.0.9

* Have you experienced this same bug with .NET Framework?: 
No

**Problem description:**
I switched my WinForms App from .NET 4.71 to .NET 5 and get the following crashes from my user's machine:
```
System.Runtime.InteropServices.ExternalException: A generic error occurred in GDI+.
Source: System.Drawing.Common
    -- get_ErrorCode = -2147467259 (0x80004005)   [int]
    -- get_HResult = -2147467259 (0x80004005)   [int]
 at [System.Drawing.Common] System.Drawing.Graphics void CheckErrorStatus(int )
 at [System.Drawing.Common] System.Drawing.Graphics void FillRectangle(System.Drawing.Brush ,System.Drawing.Rectangle )
```

From the stack trace, I can see that this crash almost always originates from showing a messagebox (drawing the background). Since I'm not able to reproduce the crash myself, I reached out to my users and I was at least able to find a common pattern:
- Many users told me that the crash happens when the machine has been in sleep mode and wakes up. My app would show a message box when a network connection is interrupted, so this makes sense.
- Some users reported that the crash happens when Windows tries to shutdown/reboot while my app is running.
- Some users reported that the crash happens when my app is set to auto-run and shows the first dialog (password prompt). In this case it seems that the app has been started while the login screen is still shown.

Judging from the user reports, I guess this crash mostly happens when the "desktop" is "hidden" or not shown (like lock screen when the machine enters sleep mode, login screen, the screen shown when you shutdown/reboot Windows).

I'm not sure how to further troubleshoot the issue or resolve the issue.

**Minimal repro:**
Unfortunately I'm not able to reliably reproduce the issue myself.

**Comment by StefanKoell** [comment]
@RussKie If I can get a hand on a special build with this attribute, I could roll that out to my users in order to find out if it resolves the issue. Unfortunately I'm not able to provide a stand alone repro. I'm not sure what other condition must be met to get the crash. I haven't been able to reproduce the crash on my machines.

**Comment by RussKie** [comment]
You can try the following: 
1. clone [release/5.0 branch](https://github.com/dotnet/runtime/tree/release/5.0), 
2. change the code, 
3. build ([instructions](https://github.com/dotnet/runtime/blob/main/docs/workflow/building/libraries/README.md#quick-start)), and then 
4. use your private binary.

**Comment by kirsan31** [comment]
> It is plausible that the missing `SetLastError` attribute is the cause, but without a stand alone repro there would be a lot of speculations, and very difficult convince the ship room to allow servicing of .NET 6.0 (RC2 is almost complete) let alone 5.0...
> 
> System.Drawing.Common belongs to dotnet/runtime, moving there.

Due to that it's related to terminal server session closing/suspending it's a _bit_ difficult to build repro :(
But calling `GetLastWin32Error` after unmanaged function without `SetLastError` flag it's a **bug** any way...

**Comment by StefanKoell** [comment]
@RussKie I tried what you suggested and even though some errors appear when I build the System.Drawing.Common solution, I do get an assembly in the net5.0-Release folder in the artifacts. Unfortunately, replacing this dll with the one I ship my product makes the app faulty. I can briefly see the process starting but it then disappears with no logs in the event log.

I also adapted the deps.json file in my app directory to pickup the correct file version from the new build (which is oddly 42.42.42.42424).

How can I include/ship my app with the manually compiled version?

Also, regarding the SetLastError flag: I checked the docs here:
https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getlastwin32error?view=net-5.0

In the Remarks section I can find the following:
There is a difference in the behavior of the GetLastWin32Error method on .NET Core and .NET Framework when DllImportAttribute.SetLastError is true. **On .NET Framework, the GetLastWin32Error method can retain error information from one P/Invoke call to the next. On .NET Core, error information is cleared before P/Invoke call, and the GetLastWin32Error represents only error information from the last method call.**

I'm not sure I fully understand the implication but there seems to be a related issue (quite fresh):
https://github.com/dotnet/runtime/issues/51600

So is the missing SetLastError in the DllImport attribute really a bug?

**Comment by kirsan31** [comment]
> In the Remarks section I can find the following:
> There is a difference in the behavior of the GetLastWin32Error method on .NET Core and .NET Framework when DllImportAttribute.SetLastError is true. **On .NET Framework, the GetLastWin32Error method can retain error information from one P/Invoke call to the next. On .NET Core, error information is cleared before P/Invoke call, and the GetLastWin32Error represents only error information from the last method call.**
> 
> I'm not sure I fully understand the implication but there seems to be a related issue (quite fresh):
> #51600
> 
> So is the missing SetLastError in the DllImport attribute really a bug?

Key moment here is: _when DllImportAttribute.SetLastError **is true**._

**Comment by RussKie** [comment]
@StefanKoell I'm afraid I don't know enough about the runtime, area owners should be able to help. (Expect some delays due public holidays in the US).
I think @safern owns this library but he's OOF. @ericstj is there someone else who can help?

@kirsan31 this is a perfect Q for @AaronRobinsonMSFT :)

**Comment by AaronRobinsonMSFT** [comment]
> In .net I see no SetLastError=true (on all of imports there)...

That seems like the issue here. 

> For us important part is Marshal.GetLastWin32Error:
> > You can use this method to obtain error codes only if you apply the System.Runtime.InteropServices.DllImportAttribute to the method signature and set the DllImportAttribute.SetLastError field to true. The process for this varies depending upon the source language used: C# and C++ are false by default, but the Declare statement in Visual Basic is true.

Right. If error codes are expected from a P/Invoke, the `SetLastError` should be set to `true`.

It appears the change here was done quite some time ago and to a different mechanism: https://github.com/dotnet/corefx/pull/39275.

/cc @elinor-fung @JeremyKuhne

**Comment by JeremyKuhne** [comment]
Sorry, I've been out for the holiday weekend.

Note that GDI+ does _not_ document that it sets last error. Looking through the sources it could, theoretically, set last error when it gets an HRESULT from some APIs that return an HRESULT with the Win32 facility code. That shows up in the bitmap and filter code. Outside of that, the underlying Win32 APIs that get called certainly can and do set last error.

The simplest thing here would be to plaster `SetLastError=true` on the P/Invokes (ugh). If we're talking minimum change (for 6) we could just P/Invoke directly to `GetLastError` here. The risk should be minimal for P/Invoking directly to `GetLastError`.

@maryamariyan @michaelgsharp @safern @tarekgh

**Comment by StefanKoell** [comment]
I'm really not an expert here but I don't understand why the SetLastError flags have been removed in 2019. It seems to be the root cause for those crashes as (at least from what I can tell) there's no other difference between the .NET Framework 4.71 implementation and the .NET Core one. As a non-expert (without knowing the reason for the SetLastError removal) this seems to be a regression which is causing real issues in production apps. I get A LOT of these crash reports and my users are quite annoyed by this "bug".

I know it's hard since there's no reliable repro case but I have one user who can reproduce the issue most of the time (9 out of 10 times). If I could get my patched DLL to work in my production code, I could at least report back that this would fix the issue.

Quite frankly, unless there's not a reasonable explanation why the SetLastError has been removed, it should be added again to be on par with the .NET Framework implementation.

Can anyone here provide the reason why they have been removed?

**Comment by ericstj** [comment]
> Can anyone here provide the reason why they have been removed?

My best guess is that it just got missed when folks converted to dynamic loading, then didn't get brought back when dynamic loading was removed in favor of [SetDllImportResolver](https://github.com/JeremyKuhne/corefx/blob/be1c71a81a7c51f1893480d21e2370c48614e3de/src/System.Drawing.Common/src/System/Drawing/LibraryResolver.cs#L16).  Perhaps the GetProcAddress behavior in pre-.NET 5 was less susceptible to runtime interruption that clobbered GetLastError.

I think @JeremyKuhne called out why it might not be right to call SetLastError (and for that matter GetLastError as well) since GDI+ doesn't document that they actually set it, you could be reading an error from some indirect call.  Imagine GDI+ does something as part of it's normal execution, say it checks for the existence of a registry key, the registry key doesn't exist, which GDI+ sees and deems acceptable, but then fails for some other totally unrelated reason.  The error in the last-error slot would be the Win32 error from the registry API which isn't relevant. If an API doesn't document that it sets last error, then you are effectively depending on undocumented side-effects of that API.

Judging from the System.Drawing code and comments on CheckErrorStatus, that seems to be exactly what System.Drawing is trying to do: draw some conclusions about why GDI+ returned a particular error, when it's own status wasn't granular enough.  A better solution here would be to ask GDI+ to return a specific status that means it failed because 
https://github.com/dotnet/runtime/blob/208e377a5329ad6eb1db5e5fb9d4590fa50beadd/src/libraries/System.Drawing.Common/src/System/Drawing/Graphics.cs#L2410
Judging on the *specificity* of that comment, I can see why they might be reluctant to do that...

> The simplest thing here would be to plaster SetLastError=true on the P/Invokes (ugh). If we're talking minimum change (for 6) we could just P/Invoke directly to GetLastError here. The risk should be minimal for P/Invoking directly to GetLastError.

P/Invokng to GetLastError doesn't help.  The purpose of putting SetLastError on a PInvoke is to tell the runtime to store the value of the last error immediately upon return from the PInvoke.  Otherwise the runtime might do somethings itself (EG: JIT does something that sets last error, GC, etc) that clears the value before executing the library code that reads it.  At that point, calling Win32 GLE won't give you the right result.  I believe the only way to get the right result is to have interop store it (@AaronRobinsonMSFT could confirm).

Adding back SetLastError would restore the .NETFramework behavior, but might cost some in perf.  @JeremyKuhne since you made changes in this area for WinForms perf and WinForms team owns this codebase on .NETFramework, can you make a reccomendation?  I see 2 options: replace the usage of GetLastError with something that accomplishes the goal without relying on undocumented behavior (perhaps just checking SM_REMOTESESSION), or adding back SetLastError everywhere.

**Comment by kirsan31** [comment]
@ericstj 

> replace the usage of GetLastError with something that accomplishes the goal without relying on undocumented behavior (perhaps just checking SM_REMOTESESSION)

Judging by the comment (and code) in 'CheckErrorStatus', `SM_REMOTESESSION` is the only half of the problem, and at the same time with `SM_REMOTESESSION` we must have no `Win32Error`:
https://github.com/dotnet/runtime/blob/208e377a5329ad6eb1db5e5fb9d4590fa50beadd/src/libraries/System.Drawing.Common/src/System/Drawing/Graphics.cs#L2430-L2432

**Comment by AaronRobinsonMSFT** [comment]
> At that point, calling Win32 GLE won't give you the right result. I believe the only way to get the right result is to have interop store it (@AaronRobinsonMSFT could confirm).

Correct. Setting `SetLastError=true` and then calling [`Marshal.GetLastPInvokeError()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.getlastpinvokeerror?view=net-6.0) or prior to .NET 6,  `Marshal.GetLastWin32Error()`. In .NET 6 we also added [`Marshal.GetLastSystemError`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.getlastsystemerror?view=net-6.0) which is a straight call to `GetLastError` or `errno`, but doesn't handle the issues that @ericstj is alluding to above.

>Adding back SetLastError would restore the .NETFramework behavior, but might cost some in perf. @JeremyKuhne since you made changes in this area for WinForms perf and WinForms team owns this codebase on .NETFramework, can you make a reccomendation?

The performance impact is that the P/Invoke can't be inlined if `SetLastError` is set to `true`. @JeremyKuhne did a great job on rooting out many of the places where this was potentially possible but it may not be appropriate in this case.

**Comment by JeremyKuhne** [comment]
@AaronRobinsonMSFT, @ericstj Are we really likely to get the thread's error status reset before we return? Particularly for blittable calls? Is this something we can avoid with function pointers in the future?

The bigger performance issue here was the dynamic wrapping of GDI+/libgdiplus delegates. Maybe a middle ground is to add them back for anything that draws or works with an image. Matrix and color stuff shouldn't need this (for example).

I've been thinking about what it would mean to just ignore on `SM_REMOTESESSION`. I'm a bit worried about suppressing important errors when on a remote session. The other checks ( `ERROR_ACCESS_DENIED`, `ERROR_PROC_NOT_FOUND`) look like they might not be important when **not** on a remote session given that we've shipped without the last error stuff. 

Another interesting thing is this here: [Detecting the Remote Desktop Services environment](https://docs.microsoft.com/en-us/windows/win32/termserv/detecting-the-terminal-services-environment)  I wonder if were still leaving ourselves open to these sorts of failures here even if we put some of this back?

**Comment by AaronRobinsonMSFT** [comment]
>@AaronRobinsonMSFT, @ericstj Are we really likely to get the thread's error status reset before we return? Particularly for blittable calls?

Any number of issues can occur, but particularly a GC &ndash; which can occur anywhere in the IL stub.

> Is this something we can avoid with function pointers in the future?

I guess, but why not simply add it back? What is the concern here? P/Invokes are much better than the delegate approach &ndash; at least from a maintenance perspective.

**Comment by kirsan31** [comment]
> Another interesting thing is this here: [Detecting the Remote Desktop Services environment](https://docs.microsoft.com/en-us/windows/win32/termserv/detecting-the-terminal-services-environment) I wonder if were still leaving ourselves open to these sorts of failures here even if we put some of this back?

Yes, really interesting note...

**Comment by AaronRobinsonMSFT** [comment]
Looking at the example above:

```csharp
[DllImport(LibraryName, ExactSpelling = true)]
internal static extern int GdipFillRectangleI(HandleRef graphics, HandleRef brush, int x, int y, int width, int height);
```

I believe that `HandleRef` makes the P/Invoke so it can't be inlined anyways since it needs special marshalling. Which means adding the `SetLastError=true` isn't going to change a lot with respect to the existence of an IL stub or the performance concerns on start up.

There appears to be a non-zero amount of impact given the GDI+ APIs' long history and implicit .NET usage semantics that have been built-up over time &ndash; this is the burden of any interop.

**Comment by JeremyKuhne** [comment]
> Any number of issues can occur, but particularly a GC – which can occur anywhere in the IL stub.

On the same thread?

> I guess, but why not simply add it back?

More of a curiosity. It would be good to get to a place where we have the interop code in sources rather than generating it at runtime- wondering if we can leverage that for this part of things.

**Comment by JeremyKuhne** [comment]
> I believe that HandleRef makes the P/Invoke so it can't be inlined anyways since it needs special marshalling.

Yeah, I'd like to make this stuff explicit. We often use this incorrectly (tying to the wrong owner for example). I use an interface in WinForms to help encourage the right usage (`IHandle`).

**Comment by AaronRobinsonMSFT** [comment]
>More of a curiosity. It would be good to get to a place where we have the interop code in sources rather than generating it at runtime- wondering if we can leverage that for this part of things.

That is possible and something that AOT scenarios will benefit from but doesn't change much with the current impact described in this issue. Either the IL stub does the check or someone generates the stub in C# and then it does the check &ndash; either way it seems like it should happen.

Perhaps I am missing how this is designed so apologies for that ignorance. Let me speak only to the interop correctness and performance perspective and leave the rest to the owners of this API. There is no fundamental performance impact here for non-inlinable P/Invokes by adding `SetLastError=true` and in order to properly manage APIs setting of that across calls it should be set. Since there are places where `Marshal.GetLastWin32Error` is being checked across a P/Invoke the `SetLastError` should be set to `true`. If the P/Invoke is inlinable but now needs `SetLastError=true` then in .NET 6 we can do something to make it AOT friendly (i.e., defined in source).

**Comment by ericstj** [comment]
It seems like it's viable to put back SetLastError on some or all of the methods (we'd need some help from @JeremyKuhne if you'd want to subset them).  I think I can buy this as a safe fix that meets the bar (fixing a regression between 3.x ->5.0)

>I wonder if were still leaving ourselves open to these sorts of failures here even if we put some of this back?

Do we think it was ever the right thing to do to swallow these errors from GDI+?  Don't they represent real failures of the underlying API?  All the call stacks above look to me like they would have side-effects, so an error from the API means it didn't do the job we asked it to.  Is it feasible to tell callers they should be catching these instead?  Basically remove the special case.  I know we've reached here accidentally, but is it a better place? (despite being different than .NETFramework / .NETCore 3.1)  I guess this depends on how "lossy" other drawing API is, my impression is that it shouldn't be...

**Comment by StefanKoell** [comment]
> Is it feasible to tell callers they should be catching these instead?

I'm not sure if this is causing more trouble. I barely call these functions myself. I think most calls come from the WinForms framework and while it may be feasible to safeguard all the WinForms calls (which is probably a heck of a project itself), what about all the 3rd party libraries (DevExpress and similar). They will all suddenly "break" unless they also put their safeguards in...

**Comment by kirsan31** [comment]
@ericstj so what's decision about servicing this for 3.1 / 5?

**Comment by StefanKoell** [comment]
I'm too interested in a .NET 5 fix as my app has already been rolled out and I'm not quite sure how soon I can upgrade to .NET 6 after it has been released.

**Comment by FilipToth** [comment]
I had the same thing happen to me when I was trying to save an image into a folder where I didn’t have write privileges. It would be nice to be a bit more specific with the exception message.

**Comment by RussKie** [comment]
https://github.com/gitextensions/gitextensions/issues/9567 reports a weird issue (and I observed this myself as well), which may be related to this as well.

**Comment by ericstj** [comment]
> @ericstj so what's decision about servicing this for 3.1 / 5?

First things first we need to decide what we do for 6.0.  For that I'm still waiting for to hear back from @dotnet/dotnet-winforms / @JeremyKuhne around the questions here: https://github.com/dotnet/runtime/issues/58741#issuecomment-917176972.  I put up a PR with what I believe is a minimal fix if we want to bring this back: https://github.com/dotnet/runtime/pull/59096

**Comment by JeremyKuhne** [comment]
@ericstj I think the right immediate answer is to put the attributes back on drawing APIs as your change does.

**Comment by ericstj** [comment]
We made this fix to 6.0 and 7.0.  Will leave this open to port for servicing.

**Comment by safern** [comment]
I will close since the PR to `release/6.0` is merged. Thanks @ericstj

**Comment by kirsan31** [comment]
Please explain in the end - whether the service will be in 5.0 or not?
Thanks.

**Comment by safern** [comment]
I'll leave that decision to @ericstj.

**Comment by ericstj** [comment]
This was fixed in 5.0 and 3.1 in the November 2021 updates.
